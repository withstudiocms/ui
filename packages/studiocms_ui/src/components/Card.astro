---
import type { HTMLTag, Polymorphic } from 'astro/types';

/**
 * Props for the card component.
 */
type Props<As extends HTMLTag = 'div'> = Omit<Polymorphic<{ as: As }>, 'as'> & {
  /**
   * The polymorphic component to render the card as. Defaults to `div`.
   */
	as?: As;
  /**
   * Whether the card should be full width. Defaults to `false`.
   */
	fullWidth?: boolean;
  /**
   * Whether the card should be full height. Defaults to `false`.
   */
	fullHeight?: boolean;
  /**
   * The variant of the card. Defaults to `default`.
   */
  variant?: 'default' | 'filled';
};

const { 
  as: As = 'div',
  fullWidth,
  fullHeight,
  variant = 'default',
  ...props
} = Astro.props;
---
<As class="sui-card" class:list={[fullWidth && "full-w", fullHeight && "full-h", variant]} {...props}>
  <div class="sui-card-header">
    <slot name="header" />
  </div>
  <div class="sui-card-body">
    <slot />
  </div>
  <div class="sui-card-footer">
    <slot name="footer" />
  </div>
</As>
<style>
  .sui-card {
    border: 1px solid hsl(var(--border));
    background-color: hsl(var(--background-base));
    border-radius: .5rem;
    width: fit-content;
    height: fit-content;
  }

  .sui-card.filled {
    background-color: hsl(var(--background-step-3));
    border: none;
  }

  .sui-card.full-w {
    width: 100%;
  }

  .sui-card.full-h {
    height: 100%;
  }

  .sui-card-header:has(*) {
    padding: 1rem 1rem 0rem 1rem;
    * {
      margin: 0;
    }
  }

  .sui-card-body {
    padding: 1rem;
  }

  .sui-card-footer:has(*) {
    border-top: 1px solid hsl(var(--border));
    padding: 1rem;
  }

  .filled .sui-card-footer {
    border: none;
  }

  @media screen and (max-width: 840px) {
    .sui-card {
      width: 100%;
    }
  }
</style>
