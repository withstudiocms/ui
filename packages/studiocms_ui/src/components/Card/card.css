.sui-card {
	border: 1px solid hsl(var(--border));
	background-color: hsl(var(--background-base));
	border-radius: var(--radius-md);
	width: fit-content;
	height: fit-content;
}

.sui-card.filled {
	background-color: hsl(var(--background-step-3));
	border: none;
}

.sui-card.full-w {
	width: 100%;
}

.sui-card.full-h {
	height: 100%;
}

.sui-card-header:has(*) {
	padding: 1rem 1rem 0rem 1rem;
	* {
		margin: 0;
	}
}

.sui-card-body {
	padding: 1rem;
}

.sui-card-footer:has(*) {
	border-top: 1px solid hsl(var(--border));
	padding: 1rem;
}

.sui-card-footer:not(:has(*)) {
	display: none;
}

.filled .sui-card-footer {
	border: none;
}

@media screen and (max-width: 840px) {
	.sui-card {
		width: 100%;
	}
}
