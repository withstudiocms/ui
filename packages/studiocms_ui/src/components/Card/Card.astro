---
import type { HTMLTag, Polymorphic } from 'astro/types';
import './card.css';

/**
 * Props for the card component.
 */
type Props<As extends HTMLTag = 'div'> = Omit<Polymorphic<{ as: As }>, 'as'> & {
	/**
	 * The polymorphic component to render the card as. Defaults to `div`.
	 */
	as?: As;
	/**
	 * Whether the card should be full width. Defaults to `false`.
	 */
	fullWidth?: boolean;
	/**
	 * Whether the card should be full height. Defaults to `false`.
	 */
	fullHeight?: boolean;
	/**
	 * The variant of the card. Defaults to `default`.
	 */
	variant?: 'default' | 'filled';
};

const { as: As = 'div', fullWidth, fullHeight, variant = 'default', ...props } = Astro.props;
---
<As class="sui-card" class:list={[fullWidth && "full-w", fullHeight && "full-h", variant]} {...props}>
  <div class="sui-card-header">
    <slot name="header" />
  </div>
  <div class="sui-card-body">
    <slot />
  </div>
  <div class="sui-card-footer">
    <slot name="footer" />
  </div>
</As>
