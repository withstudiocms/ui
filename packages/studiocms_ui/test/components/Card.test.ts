import { describe, expect } from 'vitest';
import Card from '../../src/components/Card/Card.astro';
import { test } from '../fixtures/vitest/AstroContainer';

const lorem =
	'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.';

describe('Card Component', () => {
	test('renders Card component correctly', async ({ renderComponent }) => {
		const result = await renderComponent(Card, 'Card', { slots: { default: lorem } });
		expect(result).toMatchSnapshot();
	});

	test.for([
		{ fullWidth: true },
		{ fullHeight: true },
		{ variant: 'default' },
		{ variant: 'filled' },
	])('renders Card with props %s', async (props, { renderComponent }) => {
		const result = await renderComponent(Card, 'Card', {
			props,
			slots: { default: lorem },
		});
		expect(result).toMatchSnapshot();
	});

	test.for([
		{ name: 'default', content: 'Default Slot' },
		{ name: 'header', content: 'Header Slot' },
		{ name: 'footer', content: 'Footer Slot' },
	])('renders Card with %s slot', async (slot, { renderComponent }) => {
		const result = await renderComponent(Card, 'Card', {
			slots: { [slot.name]: slot.content, default: lorem },
		});
		expect(result).toMatchSnapshot();
	});
});
