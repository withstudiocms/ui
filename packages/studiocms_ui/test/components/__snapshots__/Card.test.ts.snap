// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card Component > renders Card component correctly 1`] = `"<div class="sui-card default"> <div class="sui-card-header">  </div> <div class="sui-card-body"> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. </div> <div class="sui-card-footer">  </div> </div>"`;

exports[`Card Component > renders Card with { name: 'default', content: 'Default Slot' } slot 1`] = `"<div class="sui-card default"> <div class="sui-card-header">  </div> <div class="sui-card-body"> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. </div> <div class="sui-card-footer">  </div> </div>"`;

exports[`Card Component > renders Card with { name: 'footer', content: 'Footer Slot' } slot 1`] = `"<div class="sui-card default"> <div class="sui-card-header">  </div> <div class="sui-card-body"> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. </div> <div class="sui-card-footer"> Footer Slot </div> </div>"`;

exports[`Card Component > renders Card with { name: 'header', content: 'Header Slot' } slot 1`] = `"<div class="sui-card default"> <div class="sui-card-header"> Header Slot </div> <div class="sui-card-body"> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. </div> <div class="sui-card-footer">  </div> </div>"`;

exports[`Card Component > renders Card with props { fullHeight: true } 1`] = `"<div class="sui-card full-h default"> <div class="sui-card-header">  </div> <div class="sui-card-body"> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. </div> <div class="sui-card-footer">  </div> </div>"`;

exports[`Card Component > renders Card with props { fullWidth: true } 1`] = `"<div class="sui-card full-w default"> <div class="sui-card-header">  </div> <div class="sui-card-body"> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. </div> <div class="sui-card-footer">  </div> </div>"`;

exports[`Card Component > renders Card with props { variant: 'default' } 1`] = `"<div class="sui-card default"> <div class="sui-card-header">  </div> <div class="sui-card-body"> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. </div> <div class="sui-card-footer">  </div> </div>"`;

exports[`Card Component > renders Card with props { variant: 'filled' } 1`] = `"<div class="sui-card filled"> <div class="sui-card-header">  </div> <div class="sui-card-body"> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. </div> <div class="sui-card-footer">  </div> </div>"`;
